sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -313,15 +314,18 @@ def process_bundle(self, request, instruction_id):
       with bundle_processor.state_handler.process_instruction_id(
           instruction_id):
         with self.maybe_profile(instruction_id):
-          delayed_applications = bundle_processor.process_bundle(instruction_id)
+          delayed_applications, requests_finalization = (
+            bundle_processor.process_bundle(instruction_id))
           response = beam_fn_api_pb2.InstructionResponse(
               instruction_id=instruction_id,
               process_bundle=beam_fn_api_pb2.ProcessBundleResponse(
                   residual_roots=delayed_applications,
                   metrics=bundle_processor.metrics(),
-                  monitoring_infos=bundle_processor.monitoring_infos()))
-      # TODO(boyuanz): Don't release here if finalize is needed.
-      self.bundle_processor_cache.release(instruction_id)
+                  monitoring_infos=bundle_processor.monitoring_infos(),
+                  requires_finalization=requests_finalization))
+      # Don't release here if finalize is needed.
+      if not bundle_processor.requires_finalization():
```

Comment:
For consistency, either re-use the `requests_finalization` variable here, or use the `bundle_processor.requires_finalization()` method above instead of a double return value.

Reply:
fixed here: https://github.com/apache/beam/pull/7937/commits/03e168102db43873e1170804df4e5f3afe6d0e8d